### PR TITLE
Count VACOLS requests in tests and simulate latency locally

### PR DESCRIPTION
**Pull request by lowellrex**
This PR improves our ability to understand VACOLS' impact on Caseflow by giving us the ability to:
1. Count requests to FACOLS in our tests
2. Simulate network latency to VACOLS

# How do we do this?

To count requests to FACOLS in our tests add `DatabaseRequestCounter.enable` in a before block and then assert that the number of requests we make while running the subject code block matches our expectation:
```ruby
before { DatabaseRequestCounter.enable }
after { DatabaseRequestCounter.disable }

it "only makes 4 requests to VACOLS" do
  subject
  expect(DatabaseRequestCounter.get_counter(:vacols)).to eq(4)
end
```

To simulate VACOLS latency when running Caseflow in our local development environments we set the `SIMULATE_VACOLS_LATENCY` environment variable (and optionally `VACOLS_DELAY_MS`) when running our Rails server.
```sh
caseflow $ SIMULATE_VACOLS_LATENCY=true REACT_ON_RAILS_ENV=HOT bundle exec rails s -p 3000
# The Rails server runs with the default latency of 80ms total above.
# In order to simulate greater latency we can set that as well:
caseflow $ VACOLS_DELAY_MS=500 SIMULATE_VACOLS_LATENCY=true REACT_ON_RAILS_ENV=HOT bundle exec rails s -p 3000
```

# Why do we do this?

### Count requests to FACOLS in our tests

Latency from the Caseflow application to VACOLS averages 80 milliseconds (source: [Datadog](https://app.datadoghq.com/dashboard/54w-efy-r5d/va-systems?fullscreen_widget=399796003)) but often jumps to 100 milliseconds or more. Many requests to Caseflow trigger dozens of requests to VACOLS incurring a cost of seconds in network roundtrips alone (before including execution time of the queries which are not always optimized). By counting requests to FACOLS in our automated tests we can be certain that we are only making the number of requests to VACOLS that we expect to make, eliminating the possibility that code changes inadvertently increase the number of VACOLS requests and dramatically slow down responsiveness of the application in production.

### Simulate network latency to VACOLS

Simulating network latency to VACOLS when running Caseflow in our local development environments allows us to better understand the pain Caseflow users dealing with production latency. For example, visiting `BVASCASPER1`'s queue while simulating the default VACOLS latency results in a ~5 second slower page load.

~10 seconds no latency | ~15 seconds simulating latency
--- | ---
image | image

### CODE REVIEW

File: app/services/metrics_service.rb [hunk 1]
```diff
@@ -48,6 +48,7 @@ def self.record(description, service: nil, name: "unknown")
       metric = PrometheusService.send("#{service}_request_attempt_counter".to_sym)
       metric.increment(app: app, name: name)
       increment_datadog_counter("request_attempt", service, name, app)
+      LocalCounter.increment_counter(service)
```
Comment (lowellrex) [inline app/services/metrics_service.rb]: I'm not sure if this is the best place for this. At the current moment is only captures places in the code where we call `MetricsService.record` which includes a lot of VACOLS calls but misses most `ActiveRecord` requests.

File: spec/models/queue_config_spec.rb [hunk 2]
```diff
@@ -117,6 +123,9 @@
             expect(tabs[1][:tasks].pluck(:id)).to match_array(assigned_tasks.map { |t| t.id.to_s })
             expect(tabs[2][:tasks].pluck(:id)).to match_array(on_hold_tasks.map { |t| t.id.to_s })
             expect(tabs[3][:tasks].pluck(:id)).to match_array(completed_tasks.map { |t| t.id.to_s })
+
+            # We make 1 request to VACOLS in order to eager load data for Legacy Appeals.
+            expect(LocalCounter.get_counter(:vacols)).to eq(tabs.count)
```
Comment (lowellrex) [inline spec/models/queue_config_spec.rb]: 🎉 Exposed an area to increase efficiency significantly by avoiding this call if we don't have any tasks related to `LegacyAppeal`s in a given tab.

Comment (lowellrex) [reply spec/models/queue_config_spec.rb]: Ticketed as #11719

File: app/services/local_counter.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
+
+class LocalCounter
```
Comment (pkarman) [inline app/services/local_counter.rb]: this class name isn't very meaningful to me. How about `DBCounter::Local` or something like that? It's not clear what it counts w/o the "db"

File: spec/controllers/tasks_controller_spec.rb [hunk 4]
```diff
@@ -770,6 +778,8 @@
         expect(task["attributes"]["user_id"]).to eq(judge_user.css_id)
         expect(task["attributes"]["appeal_id"]).to eq(legacy_appeal.id)
         expect(task["attributes"]["available_actions"].size).to eq 2
+
+        expect(DatabaseRequestCounter.get_counter(:vacols)).to eq(18)
```
Comment (lowellrex) [inline spec/controllers/tasks_controller_spec.rb]: 18 VACOLS requests are as follows:
```sql
1 # VACOLS::Case.find
SELECT  "BRIEFF".* FROM "BRIEFF" WHERE "BRIEFF"."BFKEY" = :a1 FETCH FIRST :a2 ROWS ONLY

2 # VACOLS::Correspondent.find
SELECT "CORRES".* FROM "CORRES" WHERE "CORRES"."STAFKEY" = '1'

3 # VACOLS::CaseIssue.find
SELECT "ISSUES".* FROM "ISSUES" WHERE "ISSUES"."ISSKEY" = '1'

4 # VACOLS::Folder.find
SELECT "FOLDER".* FROM "FOLDER" WHERE "FOLDER"."TICKNUM" = '1'

5 # VACOLS::CaseAssignment.select_tasks
SELECT brieff.bfkey as vacols_id, brieff.bfcorlid as vbms_id, brieff.bfd19 as docket_date, brieff.bfdloout as assigned_to_location_date, decass.deassign as assigned_to_attorney_date, decass.dereceive as reassigned_to_judge_date, decass.decomp as date_completed, decass.dedocid as document_id, decass.deprod as work_product, decass.deatty as attorney_id, s1.snamef as added_by_first_name, s1.snamemi as added_by_middle_name, s1.snamel as added_by_last_name, s1.sdomainid as added_by_css_id, decass.dedeadline as date_due, decass.deadtim as created_at, decass.demdtim as updated_at, folder.tinum as docket_number, s3.snamef as assigned_by_first_name, s3.snamel as assigned_by_last_name, s3.sdomainid as assigned_by_css_id, s2.sdomainid as assigned_to_css_id, s4.snamef as written_by_first_name, s4.snamel as written_by_last_name FROM "BRIEFF" LEFT JOIN decass
            ON brieff.bfkey = decass.defolder
          LEFT JOIN staff s1
            ON decass.deadusr = s1.slogid
          JOIN staff s2
            ON brieff.bfcurloc = s2.slogid
          JOIN folder
            ON brieff.bfkey = folder.ticknum
          LEFT JOIN staff s3
            ON decass.demdusr = s3.slogid
          LEFT JOIN staff s4
            ON decass.deatty = s4.sattyid WHERE (brieff.bfkey = '1')

6 # VACOLS::Case.find
SELECT  "BRIEFF".* FROM "BRIEFF" WHERE "BRIEFF"."BFKEY" = :a1 FETCH FIRST :a2 ROWS ONLY

7 # VACOLS::Folder.find
SELECT "FOLDER".* FROM "FOLDER" WHERE "FOLDER"."TICKNUM" = '1'

8 # VACOLS::Correspondent.find
SELECT "CORRES".* FROM "CORRES" WHERE "CORRES"."STAFKEY" = '1'

9 # VACOLS::CaseIssue.find
SELECT "ISSUES".* FROM "ISSUES" WHERE "ISSUES"."ISSKEY" = '1'

10 # VACOLS::Case.aod
SELECT "BRIEFF"."BFKEY", aod FROM "BRIEFF" left join (
      select BRIEFF.BFKEY AODKEY,
        (case when (nvl(AOD_DIARIES.CNT, 0) + nvl(AOD_HEARINGS.CNT, 0)) > 0 then 1 else 0 end) AOD
      from BRIEFF

      left join (
        select TSKTKNM, count(*) CNT
        from ASSIGN
        where TSKACTCD in ('B', 'B1', 'B2')
        group by TSKTKNM
      ) AOD_DIARIES
      on AOD_DIARIES.TSKTKNM = BRIEFF.BFKEY
      left join (
        select FOLDER_NR, count(*) CNT
        from HEARSCHED
        where HEARING_TYPE IN ('C', 'T', 'V')
          AND AOD IN ('G', 'Y')
        group by FOLDER_NR
      ) AOD_HEARINGS
      on AOD_HEARINGS.FOLDER_NR = BRIEFF.BFKEY
    )
    on AODKEY = BFKEY WHERE "BRIEFF"."BFKEY" = '1'

11 # VACOLS::CaseHearing.select_hearings
SELECT VACOLS_TEST.HEARING_VENUE(vdkey) as hearing_venue, "HEARSCHED"."HEARING_DISP", "HEARSCHED"."HEARING_PKSEQ", "HEARSCHED"."HEARING_DATE", "HEARSCHED"."HEARING_TYPE", "HEARSCHED"."NOTES1", "HEARSCHED"."FOLDER_NR", "HEARSCHED"."VDKEY", "HEARSCHED"."AOD", "HEARSCHED"."HOLDDAYS", "HEARSCHED"."TRANREQ", "HEARSCHED"."TRANSENT", "HEARSCHED"."REPNAME", "HEARSCHED"."ADDON", "HEARSCHED"."BOARD_MEMBER", "HEARSCHED"."MDUSER", "HEARSCHED"."MDTIME", "SATTYID", "BFREGOFF", "BFSO", "BFCORKEY", "BFDDEC", "BFDC", "HEARSCHED"."ROOM", "HEARSCHED"."VDBVAPOC", staff.sdomainid as css_id, brieff.bfac, staff.slogid, corres.saddrst1, corres.saddrst2, corres.saddrcty, corres.saddrstt, corres.saddrcnty, corres.saddrzip, corres.snamef, corres.snamemi, corres.snamel, corres.sspare1, corres.sspare2, corres.sspare3, folder.tinum FROM "HEARSCHED" left outer join staff on staff.sattyid = board_member left outer join brieff on brieff.bfkey = folder_nr left outer join folder on folder.ticknum = brieff.bfkey left outer join corres on corres.stafkey = bfcorkey WHERE "HEARSCHED"."HEARING_TYPE" IN ('V', 'T', 'C') AND "HEARSCHED"."FOLDER_NR" = '1'

12 # VACOLS::CaseIssue.descriptions
      select
        ISSUES.ISSKEY,
        ISSUES.ISSSEQ,
        ISSUES.ISSDC,
        ISSUES.ISSDCLS,
        ISSUES.ISSDESC,
        ISSUES.ISSPROG,
        ISSUES.ISSCODE,
        ISSUES.ISSLEV1,
        ISSUES.ISSLEV2,
        ISSUES.ISSLEV3,
        ISSREF.PROG_DESC ISSPROG_LABEL,
        ISSREF.ISS_DESC ISSCODE_LABEL,
        case when ISSUES.ISSLEV1 is not null then
          case when ISSREF.LEV1_CODE = '##' then
            VFTYPES.FTDESC else ISSREF.LEV1_DESC
          end
        end ISSLEV1_LABEL,
        case when ISSUES.ISSLEV2 is not null then
          case when ISSREF.LEV2_CODE = '##' then
            VFTYPES.FTDESC else ISSREF.LEV2_DESC
          end
        end ISSLEV2_LABEL,
        case when ISSUES.ISSLEV3 is not null then
          case when ISSREF.LEV3_CODE = '##' then
            VFTYPES.FTDESC else ISSREF.LEV3_DESC
          end
        end ISSLEV3_LABEL

      from ISSUES

      inner join ISSREF
        on ISSUES.ISSPROG = ISSREF.PROG_CODE
        and ISSUES.ISSCODE = ISSREF.ISS_CODE
        and (ISSUES.ISSLEV1 is null
          or ISSREF.LEV1_CODE = '##'
          or ISSUES.ISSLEV1 = ISSREF.LEV1_CODE)
        and (ISSUES.ISSLEV2 is null
          or ISSREF.LEV2_CODE = '##'
          or ISSUES.ISSLEV2 = ISSREF.LEV2_CODE)
        and (ISSUES.ISSLEV3 is null
          or ISSREF.LEV3_CODE = '##'
          or ISSUES.ISSLEV3 = ISSREF.LEV3_CODE)

      left join VFTYPES
        on VFTYPES.FTTYPE = 'DG'
        and ((ISSREF.LEV1_CODE = '##' and 'DG' || ISSUES.ISSLEV1 = VFTYPES.FTKEY)
          or (ISSREF.LEV2_CODE = '##' and 'DG' || ISSUES.ISSLEV2 = VFTYPES.FTKEY)
          or (ISSREF.LEV3_CODE = '##' and 'DG' || ISSUES.ISSLEV3 = VFTYPES.FTKEY))

      where ISSUES.ISSKEY IN ('1')

13 # VACOLS::RemandReason.find
SELECT "RMDREA".* FROM "RMDREA" WHERE "RMDREA"."RMDKEY" = '1' AND (rmdissseq IS NOT NULL)

14 # VACOLS::Case.find
SELECT "BRIEFF".* FROM "BRIEFF" WHERE "BRIEFF"."BFKEY" IN ('1', '1', '1', '1')

15 # VACOLS::Correspondent.find
SELECT "CORRES".* FROM "CORRES" WHERE "CORRES"."STAFKEY" = '1'

16 # VACOLS::CaseIssue.find
SELECT "ISSUES".* FROM "ISSUES" WHERE "ISSUES"."ISSKEY" = '1'

17 # VACOLS::Folder.find
SELECT "FOLDER".* FROM "FOLDER" WHERE "FOLDER"."TICKNUM" = '1'

18 # VACOLS::Case.aod
SELECT "BRIEFF"."BFKEY", aod FROM "BRIEFF" left join (
      select BRIEFF.BFKEY AODKEY,
        (case when (nvl(AOD_DIARIES.CNT, 0) + nvl(AOD_HEARINGS.CNT, 0)) > 0 then 1 else 0 end) AOD
      from BRIEFF

      left join (
        select TSKTKNM, count(*) CNT
        from ASSIGN
        where TSKACTCD in ('B', 'B1', 'B2')
        group by TSKTKNM
      ) AOD_DIARIES
      on AOD_DIARIES.TSKTKNM = BRIEFF.BFKEY
      left join (
        select FOLDER_NR, count(*) CNT
        from HEARSCHED
        where HEARING_TYPE IN ('C', 'T', 'V')
          AND AOD IN ('G', 'Y')
        group by FOLDER_NR
      ) AOD_HEARINGS
      on AOD_HEARINGS.FOLDER_NR = BRIEFF.BFKEY
    )
    on AODKEY = BFKEY WHERE "BRIEFF"."BFKEY" IN ('1', '1', '1', '1')
```

Determined by adding the following code in `vacols_request_spy()` to print the statements we make:
```ruby
if DatabaseRequestCounter.enabled?
  puts(DatabaseRequestCounter.get_counter(:vacols))
  puts(statement)
  puts("\n")
end
```

**Comment by lowellrex** [comment]
Just now noticed that I spelled my own name wrong on the branch name... Very disappointing

**Pull request merged**